crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -1568,51 +1574,59 @@ impl CumulocityConverter {
             }
         };
 
+        // Save the operation to the operation directory
         let ops_file = ops_dir.join(c8y_operation_name);
         create_directory_with_defaults(&*ops_dir)?;
         create_file_with_defaults(&ops_file, None)?;
 
-        let c8y_prefix = self.config.bridge_config.c8y_prefix.clone();
+        let ops_dir = ops_dir.as_std_path();
 
-        let operation = c8y_api::smartrest::operations::get_operation(
-            ops_file.as_std_path(),
-            &self.config.bridge_config,
-        )?;
-        let operations = self
-            .operations_for_device_mut(target)
-            .expect("entity should've been checked before that it's not a service");
+        let need_cloud_update = match is_child_operation_path(ops_dir) {
+            // for devices other than the main device, dynamic update of supported operations via file events is
+            // disabled, so we have to additionally load new operations from the c8y operations for that device
+            true => self.update_operations(ops_dir)?,
 
-        let prev_operation = operations.remove_operation(&operation.name);
-        // even if the body of the operation is different, as long as it has the same name, supported operations message
-        // will be the same, so we don't need to resend
-        let need_cloud_update = prev_operation.is_none();
-        operations.add_operation(operation);
+            // for main devices new operation files are loaded dynamically as they are created, so only register one
+            // operation we need
+            false => {
+                let bridge_config = &self.config.bridge_config;
+                let operation = c8y_api::smartrest::operations::get_operation(
+                    ops_file.as_std_path(),
+                    bridge_config,
+                )?;
 
-        if need_cloud_update {
-            let topic = if is_child_operation_path(ops_dir.as_std_path()) {
-                let child_id = get_child_external_id(ops_dir.as_std_path())?;
-                let child_external_id = Self::validate_external_id(&child_id)?;
+                let current_operations = self
+                    .operations_for_device_mut(target)
+                    .expect("entity should've been checked before that it's not a service");
 
-                C8yTopic::ChildSmartRestResponse(child_external_id.into()).to_topic(&c8y_prefix)?
-            } else {
-                C8yTopic::upstream_topic(&c8y_prefix)
-            };
+                let prev_operation = current_operations.remove_operation(&operation.name);
```

Comment:
In your refactoring PR, consider updating the signature of `Operations::add_operation` function to return the previous operation so that we can avoid this removal first and addition one step later.